n_queen_problem/NQueenProblem.java

```diff
@@ -1,64 +1,171 @@
+                        builder.append(EMPTY_SQUARE);
+                    }
+                }
+                builder.append(NEW_LINE);
+            }
+            return builder.toString();
+        }
     }
 
-    private static boolean solveNQueen(int[][] board, int column) {
-        int boardLength = board[0].length;
-        if (column >= boardLength)
-            return true;
+    private class NQueenIterator implements Iterator<Solution> {
 
-        for (int i = 0; i < boardLength; ++i) {
-            if (isQueenSafe(board, i, column)) {
-                board[i][column] = 1;  // Place queen on (row, column) = (i, column).
-                if (solveNQueen(board, column + 1))  // Recurse for remaining board
+        /**
+         * Queue to store the intermediate boards necessary to build all solutions.
+         */
+        private final Deque<List<Integer>> queue;
+        /**
+         * Next solution to be returned.
+         */
+        private Solution next;
+
+        /**
+         * Initializes the queue and adds an empty board to it.
+         */
+        private NQueenIterator() {
+            this.queue = new ArrayDeque<>();
+            queue.add(new ArrayList<>(size));
+        }
+
+        /**
+         * <p>Returns true is there's one more solution.</p>
+         *
+         * <p>It uses a backtracking algorithm to check for a possible solution. It removes the last board enqueued
+         * and for the next empty column it tries to place a queen in every square. If that queen doesn't attack any of the already
+         * placed queens, it queues a board with that queen. Once it removes a board with 8 queens, it stops and returns true</p>
+         */
+        @Override
+        public boolean hasNext() {
+            while (!queue.isEmpty()) {
+                List<Integer> current = queue.removeLast();
+                if (current.size() == size) {
+                    this.next = new Solution(current);
                     return true;
-                else
-                    board[i][column] = 0;
+                }
+                int column = current.size();
+                for (int row = 0; row < size; ++row) {
+                    if (valid(current, column, row)) {
+                        List<Integer> next = new ArrayList<>(current);
+                        next.add(row);
+                        queue.addLast(next);
+                    }
+                }
             }
+            return false;
+        }
+
+        @Override
+        public Solution next() {
+            return next;
+        }
+
+        /**
+         * <p>Returns true if is valid to place a queen on the square (column, row)</p>
+         *
+         * <p>A queen can be placed if it doesn't attack any of the already placed queens: there can't be a queen
+         * already in the same row, and the queen can't be collinear with any other.</p>
+         *
+         * @param board
+         * @param column
+         * @param row
```

Comment:
Please write a bit about the parameters here and at other places too.

Reply:
¡Will do, thanks!